source/Ship.cpp

```diff
@@ -3710,9 +3710,25 @@ void Ship::DoGeneration()
 
 	isDisabled = isOverheated || hull < MinimumHull() || (!crew && RequiredCrew());
 
+	double coolingEfficiency = CoolingEfficiency();
+	heat -= coolingEfficiency * attributes.Get("cooling");
 	// Update ship supply levels.
 	if(isDisabled)
+	{
 		PauseAnimation();
+		// If overheated, but otherwise not disabled, apply active cooling.
+		if(!(hull < MinimumHull() || (!crew && RequiredCrew())))
+		{
+			// If overheated, heat must be >= 100%
+			// So there is no need to check it again when handling active cooling.
+			// Still need to check for energy though.
+			if(energy >= 0.)
+			{
+				heat -= coolingEfficiency * attributes.Get("active cooling");
+				energy -= attributes.Get("cooling energy");
+			}
+		}
```

Comment:
Can you put this in a separate `if` before this one? So that you can just check for `isOverheated` instead.